File: McpAppsDotnet/server/Steps/Step6FullscreenCsp.cs
using System.ComponentModel;
using System.Text.Json;
using ModelContextProtocol.Extensions.Apps;
using ModelContextProtocol.Protocol;
using ModelContextProtocol.Server;

namespace McpAppsDotnet.Steps;

/// <summary>
/// Step 6 — Display modes + external resources via CSP.
///
/// Two new pieces of <c>_meta.ui</c> show up here, expressed as a
/// <see cref="McpUiResourceMeta"/> and attached to the <em>content</em> of the
/// <c>resources/read</c> result (and, for completeness, to the resource listing —
/// see <see cref="Resource"/>):
/// <list type="bullet">
///   <item><see cref="McpUiResourceCsp.ResourceDomains"/> (<c>csp.resourceDomains</c>)
///   — origins the View may load <c>&lt;img&gt;</c>/<c>&lt;script&gt;</c>/font/media
///   from (the <c>img-src</c>/<c>script-src</c>/… buckets). Use
///   <see cref="McpUiResourceCsp.ConnectDomains"/> for <c>fetch</c>/XHR/WebSocket
///   targets. The host's default CSP is effectively <c>default-src 'none'</c>, so
///   without this whitelist the <c>flagcdn.com</c> image is blocked.</item>
///   <item><see cref="McpUiResourceMeta.PrefersBorder"/> (<c>prefersBorder</c>) — a
///   cosmetic hint; a <em>direct</em> field of <c>_meta.ui</c> (sibling of
///   <c>csp</c>/<c>permissions</c>), not nested under any wrapper.</item>
/// </list>
/// The View asks for fullscreen via <c>app.requestDisplayMode</c> — see
/// <c>ui/src/step6-fullscreen-csp/view.ts</c>.
/// </summary>
public static class Step6FullscreenCsp
{
    private const string ResourceUri = "ui://step6-fullscreen-csp/app.html";

    private static readonly Dictionary<string, string> Countries = new()
    {
        ["at"] = "Austria",
        ["de"] = "Germany",
        ["fr"] = "France",
        ["jp"] = "Japan",
        ["us"] = "United States",
        ["br"] = "Brazil",
    };

    public static IEnumerable<McpServerTool> Tools() =>
    [
        McpApps.SetAppUi(
            McpServerTool.Create(Flag, new McpServerToolCreateOptions
            {
                Name = "step6-flag",
                Title = "Step 6 — Country Flag",
                Description = "Shows a country flag image from an external CDN inside a sandboxed iframe. " +
                              "The View can toggle fullscreen.",
            }),
            new McpUiToolMeta { ResourceUri = ResourceUri }),
    ];

    public static McpServerResource Resource(ViewStore views)
    {
        // The same McpUiResourceMeta is published on every level a host might look at.
        //
        // 1) Content level (`views.Read(..., UiMeta)`) — the _meta of the single
        //    TextResourceContents in the resources/read result. This is the one that
        //    matters: it takes precedence over listing metadata and is what a host
        //    evaluates right before it builds the sandboxed iframe.
        // 2) Listing level (`Meta` below) — shown in resources/list, so a host that
        //    pre-scans resources knows about the CSP before reading anything.
        var resource = McpServerResource.Create(
            () => views.Read(ResourceUri, "step6.html", UiMeta),
            new McpServerResourceCreateOptions
            {
                UriTemplate = ResourceUri,
                Name = "step6-fullscreen-csp-ui",
                MimeType = McpApps.HtmlMimeType,
                Description = "Step 6 — Fullscreen + CSP view",
                Meta = ViewStore.MetaObject(UiMeta),
            });

        // 3) Template level — McpApps.SetResourceUi writes ResourceTemplate.Meta,
        //    which surfaces in resources/templates/list. Careful: for a resource whose
        //    URI has no parameters (like this one) the SDK exposes only the concrete
        //    Resource, so templates/list is empty and SetResourceUi alone would never
        //    reach the wire — hence the explicit `Meta` in step 2. Verified against
        //    ModelContextProtocol 2.2.0 with curl.
        return McpApps.SetResourceUi(resource, UiMeta);
    }

    private static CallToolResult Flag(
        [Description("ISO 3166-1 alpha-2 code, e.g. 'at'")] string? code = "at")
    {
        code = (code ?? "at").ToLowerInvariant();
        var country = Countries.TryGetValue(code, out var name) ? name : code.ToUpperInvariant();
        return new CallToolResult
        {
            Content = [new TextContentBlock { Text = $"Showing flag of {country} ({code})." }],
            StructuredContent = JsonSerializer.SerializeToElement(new { country, code }),
        };
    }

    // The typed _meta.ui for this resource. Without the ResourceDomains entry the
    // View's <img> is blocked by the host's default-deny CSP.
    //
    // Using McpUiResourceMeta instead of a hand-built JsonObject also removes a whole
    // class of typos: PrefersBorder is a property of the meta object itself, so it
    // cannot accidentally end up nested under an invented `preferences` wrapper (a
    // host would silently ignore that). Other properties available here: Permissions
    // (iframe `allow` list) and Domain (dedicated origin); Csp additionally offers
    // ConnectDomains (fetch/XHR/WebSocket), FrameDomains and BaseUris.
    private static readonly McpUiResourceMeta UiMeta = new()
    {
        Csp = new McpUiResourceCsp { ResourceDomains = ["https://flagcdn.com"] },
        PrefersBorder = true,
    };
}

File: McpApps/src/server/steps/step6-fullscreen-csp.ts
import type { McpServer } from "@modelcontextprotocol/server";
import { z } from "zod";
import { APP_MIME, contentUiMeta, uiResource } from "../lib/ui-meta.js";
import type { ViewStore } from "../lib/view-store.js";

/**
 * Step 6 — Display modes + external resources via CSP.
 *
 * Two new pieces of `_meta.ui` show up here, attached at the *content* level
 * of the `resources/read` result (see `contentUiMeta` in `lib/ui-meta.ts`):
 *   - `csp.resourceDomains` — origins the View may load
 *     `<img>`/`<script>`/font/media from (the `img-src`/`script-src`/… CSP
 *     buckets). Use `csp.connectDomains` for `fetch`/XHR/WebSocket targets.
 *     The host's default CSP is effectively `default-src 'none'`, so
 *     without this allow-list the `flagcdn.com` image is blocked.
 *   - `prefersBorder` — a cosmetic hint; a *direct* field of `_meta.ui`
 *     (sibling of `csp`/`permissions`), not nested under any wrapper.
 *
 * The View asks for fullscreen via `app.requestDisplayMode` — see
 * `src/ui/step6-fullscreen-csp/view.ts`.
 */
const RESOURCE_URI = "ui://step6-fullscreen-csp/app.html";

// The .NET original accepted any string and silently produced a 404 image
// for an unknown code. Modeling the code as an enum instead lets the host
// (and the model) validate the argument before the tool is even called, so
// "unknown country code" becomes a schema-level rejection instead of a
// broken image inside the iframe.
const COUNTRY_CODES = ["at", "de", "fr", "jp", "us", "br"] as const;

const COUNTRIES: Record<(typeof COUNTRY_CODES)[number], string> = {
  at: "Austria",
  de: "Germany",
  fr: "France",
  jp: "Japan",
  us: "United States",
  br: "Brazil",
};

export function register(server: McpServer, views: ViewStore): void {
  server.registerTool(
    "step6-flag",
    {
      title: "Step 6 — Country Flag",
      description:
        "Shows a country flag image from an external CDN inside a sandboxed iframe. " + "The View can toggle fullscreen.",
      inputSchema: z.object({
        code: z.enum(COUNTRY_CODES).default("at").describe("ISO 3166-1 alpha-2 code, e.g. 'at'"),
      }),
      outputSchema: z.object({
        country: z.string(),
        code: z.string(),
      }),
      _meta: uiResource(RESOURCE_URI),
    },
    async ({ code }) => {
      const country = COUNTRIES[code];
      return {
        content: [{ type: "text", text: `Showing flag of ${country} (${code}).` }],
        structuredContent: { country, code },
      };
    },
  );

  server.registerResource(
    "step6-fullscreen-csp-ui",
    RESOURCE_URI,
    {
      title: "Step 6 — Fullscreen + CSP view",
      description: "Step 6 — Fullscreen + CSP view",
      mimeType: APP_MIME,
    },
    async (uri) =>
      views.read(
        uri.href,
        "step6-fullscreen-csp",
        // Content-level _meta on the resource read result (takes precedence
        // over any listing-level _meta). Without the resourceDomains entry
        // the View's <img> is blocked by the host's default-deny CSP.
        contentUiMeta({
          csp: { resourceDomains: ["https://flagcdn.com"] },
          prefersBorder: true,
        }),
      ),
  );
}
